### ISSUE TITLE
Maintaining binaries for older OSs

### ISSUE THREAD

**Issue by martinjvickers** [seqan/seqan#1782]
Not an issue, just a question. 

Since the update of SeqAn that now requires GLIBC 2.14 the binaries I have created no longer run on many of the production HPC environments that run Scientific Linux 6 which is GLIB 2.12 which will be unlikely to be updated.

I was just wondering what strategies people are using to provide binary support with previous OSs. e.g. compile apps against old versions of the seqan Library?

**Comment by h-2** [comment]
What binaries are you referring to? Our official binaries of the official apps, or the applications you build yourself?

What are your build options? Are you using static linking or dynamic linking?

**Comment by martinjvickers** [comment]
Hi @h-2, thank you for the swift reply.

Sorry, I should clarify. This is an application I've built for binary release.

My application is simple in that it has no other libraries beyond what seqan uses. So all I did was add ` -static` to my CMakeLists.txt file which other than that addition is identical to the readthedocs example;

```
# Add CXX flags found by find_package (SeqAn).
set (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${SEQAN_CXX_FLAGS} -static")
```

When I compile it on my Mint 16 machine all is fine and `ldd` gives the following;

```
martin@x250 ~/development/app-build/release $ ldd app-stable 
	not a dynamic executable
```

which is what I would expect. When I spin up a centos 6.8 DO droplet the binary works but when I try a 5.11 x64 droplet I get;
```
[root@centos-512mb-lon1-01 ~]# ./alfsc-stable 
FATAL: kernel too old
```

**Comment by h-2** [comment]
Ah, ok. You are doing everything the right way. 
There are some workarounds to this, as explained here:
http://stackoverflow.com/questions/11206243/how-to-build-a-static-binary-for-gnu-linux-installations-with-old-kernel

But especially if you start using dependencies like libz or openmp it might be easier to set up a virtualmachine with the linux version you are targeting and build there.

**Comment by martinjvickers** [comment]
Hi @h-2, thanks for getting back to me. 

Okay I wasn't going mad then :-) I will give that a try. I had considered the more drastic VM for the version but I didn't know if that would cause me more issues (e.g. having to have multiple target versions). My aim it to ensure my binaries work with all current Linux distros that are not EOL (https://linuxlifecycle.com/).

I've never released binaries before and the seqan documentation has some really good information about general best practice, maybe we could include some of this type of information in the readthedocs pages?

**Comment by h-2** [comment]
>  I had considered the more drastic VM for the version but I didn't know if that would cause me more issues (e.g. having to have multiple target versions). My aim it to ensure my binaries work with all current Linux distros that are not EOL (https://linuxlifecycle.com/).

I think the best approach would be to 
* compare the abovementioned linux distributions and pick the one with the oldest kernel, setup a VM
* somehow manage to get the oldest supported seqan compiler (gcc-4.9 right now) and a compatible cmake (>= 3.0 right now) running on that VM
* build static binaries on that system

> I've never released binaries before and the seqan documentation has some really good information about general best practice, maybe we could include some of this type of information in the readthedocs pages?

That's a good idea, I will try to do it for the next release.

**Comment by martinjvickers** [comment]
I'm happy to put something together from what we've discussed. You could make any corrections or add to it if you like?

**Comment by h-2** [comment]
@martinjvickers Sure, that would be great! I thought of making an article below 
http://seqan.readthedocs.io/en/develop/Infrastructure/Misc/index.html
that contains the above hints. Could you create a pull request for that?

I would also adapt
http://seqan.readthedocs.io/en/develop/Infrastructure/Manage/Releases.html
but that needs some minor changes anyway so I need to do it.

**Comment by h-2** [comment]
Moving this to 2.3.0 because 2.2.0's release is soon.

**Comment by holtgrewe** [comment]
Are you aware of https://github.com/phusion/holy-build-box ?

**Comment by martinjvickers** [comment]
@holtgrewe I've not seen this before, it looks very interesting. I will certainly take a look at this. Many thanks.

**Comment by h-2** [comment]
@holtgrewe thanks that looks promising, the only problem is, that we need gcc4.9, not 4.8 and that we also need a static bz2 lib. But maybe we can adapt the images!

**Comment by h-2** [comment]
Depends on phusion/holy-build-box/issues/2

**Comment by h-2** [comment]
@kdmurray91 cool that you worked on this! Do the images also contain bzip2?
@rrahn we should use this for the release if possible. In the medium term: can we integrate this with our docker plans / continuous deployment?

**Comment by kdmurray91** [comment]
@h-2 Yep, the images contain everything they used to, plus bzip2 and boost 1.62.0. I've pushed another test release to docker hub, so `docker pull kdmurray91/kdm-hbb-64:latest` to update.

**Comment by rrahn** [comment]
@h-2 well I agree and I think that's possible for or docker plans anyway.

**Comment by h-2** [comment]
Let's try to use it for the release packages. Infrastructure integration moved to next release...

**Comment by h-2** [comment]
@rrahn reassigning this to you, as you will be working on the deployment, ok?

**Comment by rrahn** [comment]
It is still on our plate, and we might can work on that beginning of next year. 
I will postpone it to the next release.
